### PR TITLE
Implement 0x004B6572 - vehicle rename game command

### PR DESCRIPTION
**Pull request by petergaal**

### CODE REVIEW

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+#include "../Types.hpp"
+#include "Vehicle.h"
+
+using namespace OpenLoco::Interop;
+
+namespace OpenLoco::Vehicles
+{
+    static loco_global<uint8_t, 0x009C68EA> gGameCommandExpenditureType; // premultiplied by 4
+    static loco_global<uint32_t[9], 0x011361F8> gRenameBuffer;
+    static loco_global<char[36], 0x011361F8> gRenameBufferAsString;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:16]: Hmm this is a bit cheeky defining two pointers that point to the same address.

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        static loco_global<char[512], 0x0112CC04> _stringFormatBuffer;
+
+        auto args = FormatArguments::common(vehicleHead->var_44);                     // 004B65C6-004B65CA
+        StringManager::formatString(_stringFormatBuffer, vehicleHead->var_22, &args); // 004B65D0-004B65E6
+        if (strcmp(_stringFormatBuffer, gRenameBufferAsString) == 0)                  // 004B65F1-004B6601
+        {
+            return 0; // 004B6655-004B6657
+        }
+
+        uint16_t allocationResult = 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:58]: allocatedStringId

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                return 0;
+            }
+        }
+        else
+        {
+            if ((flags & GameCommands::apply) == 0) // 004B663D-004B6640
+            {
+                return 0; // 004B6639-004B663B, 004B6655-004B6657
+            }
+            allocationResult = static_cast<uint16_t>(vehicleHead->vehicleType) + 4; // 004B6642-004B664B
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:78]: Define string id for the 4.

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        uint16_t tempAx = vehicleHead->var_22;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:81]: oldStringId

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        gGameCommandExpenditureType = static_cast<uint8_t>(ExpenditureType::TrainRunningCosts) * 4; // 004B6572
+        if (index == 1)                                                                             // 004B6579-004B657D
+        {
+            _headId_113621D = headId; // 004B657F
+        }
+
+        if ((flags & GameCommands::apply) != 0) // 004B6586-004B6589
+        {
+            int16_t arrayIndex = index - 1; // 004B658B-004B658E
+            if (arrayIndex < 0)             // 004B658F
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:32]: Make a little table for this index to offset 2, 0, 1

File: src/OpenLoco/Vehicles/RenameVehicle.cpp (hunk 3 above)
Comment (petergaal) [reply src/OpenLoco/Vehicles/RenameVehicle.cpp]: Replaced with string ID map.

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 6]
```diff
@@ -46,18 +46,20 @@ namespace OpenLoco::Vehicles
         Vehicles::VehicleHead* vehicleHead = ThingManager::get<Vehicles::VehicleHead>(vehicleHeadId); // 004B65BD-004B65C0
 
         static loco_global<char[512], 0x0112CC04> _stringFormatBuffer;
+        char renameStringBuffer[36] = "";
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:49]: Make it 37 to allow for a null at the end

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 7]
```diff
@@ -46,18 +46,20 @@ namespace OpenLoco::Vehicles
...
+        memcpy(renameStringBuffer, staticRenameBuffer, sizeof(renameStringBuffer));
 
         auto args = FormatArguments::common(vehicleHead->var_44);                     // 004B65C6-004B65CA
         StringManager::formatString(_stringFormatBuffer, vehicleHead->var_22, &args); // 004B65D0-004B65E6
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:53]: This doesn't have to be the string format buffer. Can just make it a local buffer for this function. Then we can give it a nice name like `existingVehicleName`

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        char existingVehicleName[512];
+        auto args = FormatArguments::common(vehicleHead->var_44);                     // 004B65C6-004B65CA
+        StringManager::formatString(existingVehicleName, vehicleHead->var_22, &args); // 004B65D0-004B65E6
+        if (strcmp(existingVehicleName, renameStringBuffer) == 0)                     // 004B65F1-004B6601
+        {
+            return 0; // 004B6655-004B6657
+        }
+
+        string_id allocatedStringId = 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:60]: `StringIds::empty`

File: src/OpenLoco/Vehicles/RenameVehicle.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        if (strlen(renameStringBuffer) != 0) // 004B6603-004B660A
+        {
+            allocatedStringId = StringManager::userStringAllocate(renameStringBuffer, 0); // 004B660C-004B6613
+            if (allocatedStringId == 0)                                                   // 004B6618-004B661B
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/RenameVehicle.cpp:64]: `StringIds::empty`

**Review by duncanspumpkin — APPROVED** [review]
Can you remove all of the additional location comments.

**Review by duncanspumpkin — APPROVED** [review]

**Comment by duncanspumpkin** [comment]
You'll need to rebase due to the changes made on the gamecommand caller.

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**